workflows-cli/src/linter/linter_labels.rs

```diff
@@ -0,0 +1,330 @@
+    }
+
+    fn new() -> Result<LabelChecker, String> {
+        let rules = Self::build_rules()?;
+        Ok(LabelChecker { rules })
+    }
+
+    fn validate(&self, labels: &Value, annotations: &Value) -> Result<Vec<String>, String> {
+        let mut errors = vec![];
+
+        for rule in &self.rules.critical {
+            if let Err(e) = rule.check(labels, annotations) {
+                errors.push(e);
+            }
+        }
+
+        Ok(errors)
+    }
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(tag = "type")]
+enum CriticalRule {
+    #[serde(rename = "stem")]
+    Stem(Stem),
+    #[serde(rename = "match")]
+    Match(Match),
+}
+
+#[derive(Debug, Deserialize)]
+struct Rules {
+    critical: Vec<CriticalRule>,
+}
+
+#[derive(Debug, Deserialize)]
+struct Match {
+    key: String,
+    location: Location,
+}
+
+#[derive(Debug, Deserialize)]
+struct Stem {
+    stem: String,
+    options: Vec<String>,
+    location: Location,
+}
+
+#[derive(Debug, Deserialize)]
+enum Location {
+    #[serde(rename = "annotation")]
+    Annotation,
+    #[serde(rename = "label")]
+    Label,
+}
+
+impl fmt::Display for Location {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Annotation => write!(f, "annotations"),
+            Self::Label => write!(f, "labels"),
+        }
+    }
+}
+
+trait RuleChecker {
+    fn check(&self, labels: &Value, annotations: &Value) -> Result<(), String>;
+}
+
+impl RuleChecker for CriticalRule {
+    fn check(&self, labels: &Value, annotations: &Value) -> Result<(), String> {
+        match self {
+            CriticalRule::Match(m) => m.check(labels, annotations),
+            CriticalRule::Stem(s) => s.check(labels, annotations),
+        }
+    }
+}
+
+impl RuleChecker for Match {
+    fn check(&self, labels: &Value, annotations: &Value) -> Result<(), String> {
+        let exists = match self.location {
+            Location::Annotation => annotations.get(&self.key),
+            Location::Label => labels.get(&self.key),
+        };
+        if exists.is_none() {
+            let msg = format!("Expected {} in {}", self.key, self.location);
+            return Err(msg);
+        }
+        Ok(())
+    }
+}
+
+impl RuleChecker for Stem {
+    fn check(&self, labels: &Value, annotations: &Value) -> Result<(), String> {
+        let mut entries: Vec<(String, &Value)> = vec![];
+
+        for option in &self.options {
+            let formatted_option = format!("{}-{}", self.stem, option);
+
+            let value = match self.location {
+                Location::Annotation => annotations.get(&formatted_option),
+                Location::Label => labels.get(&formatted_option),
+            };
+
+            if let Some(val) = value {
+                entries.push((formatted_option.clone(), val));
+            }
+        }
+
+        if entries.is_empty() {
+            let msg = format!("Expected {}-<GROUP> in {}", self.stem, self.location);
```

Comment:
Stem rule code is generic over more than just `science-group` but error message imples that the `leaf` of the key is a `<GROUP>`.